DDD-libsignal-protocol-java/src/main/java/org/whispersystems/libsignal/SessionCipher.java

```diff
@@ -0,0 +1,558 @@
+        try {
+            Cipher cipher;
+
+            if (version >= 3) {
+                cipher = getCipher(Cipher.DECRYPT_MODE, messageKeys.getCipherKey(), messageKeys.getIv());
+            } else {
+                cipher = getCipher(Cipher.DECRYPT_MODE, messageKeys.getCipherKey(), messageKeys.getCounter());
+            }
+
+            return cipher.doFinal(cipherText);
+        } catch (IllegalBlockSizeException | BadPaddingException e) {
+            throw new InvalidMessageException(e);
+        }
+    }
+
+
+    private byte[] getPlaintext(int version, MessageKeys messageKeys, InputStream inputStream,
+                                OutputStream outputStream) throws InvalidMessageException, IOException {
+        try {
+            Cipher cipher;
+            if (version >= 3) {
+                cipher = getCipher(Cipher.DECRYPT_MODE, messageKeys.getCipherKey(), messageKeys.getIv());
+            } else {
+                cipher = getCipher(Cipher.DECRYPT_MODE, messageKeys.getCipherKey(), messageKeys.getCounter());
+            }
+            CipherOutputStream cipherOutputStream = new CipherOutputStream(outputStream,cipher);
+
+
+            byte[] buffer = new byte[readSize];  // ReadSize was modified during testing
+            byte[] trailingBuffer = new byte[32]; // To store the last 32 bytes
+
+            int TRAILING_SIZE = 32;
+            int trailingCount = 0;
+
+            int readCount;
+            while ((readCount = inputStream.read(buffer)) > 0) {
+                if (readCount > TRAILING_SIZE) {
+                    cipherOutputStream.write(trailingBuffer, 0 , trailingCount);
+                    cipherOutputStream.write(buffer, 0, readCount - TRAILING_SIZE);
+                    System.arraycopy(buffer, 0,trailingBuffer,0, TRAILING_SIZE);
+                }
+                else {
+                    int bytesToWriteIntoOutPut = max(0,(readCount + trailingCount) - 32); // All data - 32, write to outPutStream
+                    int writeFromTrailing = min(bytesToWriteIntoOutPut, 32);        // Number of Bytes to Send to OutputStream
+                    cipherOutputStream.write(trailingBuffer, 0, writeFromTrailing);
+                    if(writeFromTrailing > 0) { //Avoid SystemCopyIssues
+                        System.arraycopy(trailingBuffer, writeFromTrailing, trailingBuffer, 0,
+                                         trailingCount - writeFromTrailing);
+                    }
+                    int trailingBufferOffset = trailingCount - bytesToWriteIntoOutPut;
+                    System.arraycopy(buffer, 0, trailingBuffer, trailingBufferOffset,  readCount);
+                    trailingCount = min(trailingCount + readCount,32);
+                }
+            }
+
+            cipherOutputStream.close();
+            return trailingBuffer;
+        }catch(Exception e){
+            throw new IOException(e.getMessage());
```

Comment:
try not to rethrow exceptions. if it isn't an IOException or InvalidMessageException, that type and rethrow with new IOException(e.getMessage(), e); to preserve the original exception.